### PR TITLE
PLANNER-2372 Re-enable ANC when not native

### PR DESCRIPTION
**Pull request by triceo**
This assumes that we can have ANC on the classpath in native. 
If that is not true, then there is currently no solution to bring ANC back.

### CODE REVIEW

**Comment by triceo** [comment]
Jenkins run native.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseUtil.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import org.drools.ancompiler.KieBaseUpdaterANC;
+import org.drools.core.base.CoreComponentsBuilder;
+import org.kie.api.KieBase;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class KieBaseUtil {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KieBaseUtil.class);
+    private static final String ANC_ENABLED_PROPERTY = "optaplanner.drools.compile_alpha_network";
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseUtil.java:28]: Let's not use a system property, but a field in ScoreDirectorFactoryConfig
That what it's consistent with SolverConfig.domainAccessType
and we can run multiple solver instances with a different setting in the same JVM.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseUtil.java:28]: It will also need fail fasts:
if the score director is not CS-D, then fail fast if that field is non-null.

**Review by ge0ffrey — CHANGES_REQUESTED** [review]
Don't use a system property please.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java [hunk 2]
```diff
@@ -156,6 +174,35 @@ public void setScoreDrlFileList(List<File> scoreDrlFileList) {
...
 
+    public Boolean getCompileDroolsAlphaNetwork() {
+        return compileDroolsAlphaNetwork;
+    }
+
+    public void setCompileDroolsAlphaNetwork(Boolean compileDroolsAlphaNetwork) {
+        this.compileDroolsAlphaNetwork = compileDroolsAlphaNetwork;
+    }
+
+    public boolean isDroolsAlphaNetworkCompilerEnabled() {
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:185]: This method is used from Quarkus/Spring boot code injection.
I don't particularly like that it is here, but it is the only reasonable way I could think of to share this code.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java [hunk 3]
```diff
@@ -180,7 +188,17 @@ private void checkMultipleScoreDirectorFactoryTypes(EasyScoreDirectorFactory eas
         }
     }
 
-    protected ConstraintStreamScoreDirectorFactory<Solution_, Score_> buildConstraintStreamScoreDirectorFactory(
+    private void enforceNoAlphaNetworkCompiler() {
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:191]: Fail-fast.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java [hunk 4]
```diff
@@ -41,27 +38,16 @@
  * @see ConstraintStreamScoreDirector
  * @see ScoreDirectorFactory
  */
-public class ConstraintStreamScoreDirectorFactory<Solution_, Score_ extends Score<Score_>>
+public abstract class AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_ extends Score<Score_>>
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java:41]: So that I can only deal with ACN in code that is actually using Drools.

**Comment by triceo** [comment]
@rsynek Please verify the changes to solver config.
@Christopher-Chianelli Please check the changes to SB/Q injection/autowiring.
@ge0ffrey Your final approval, please.

**Review by rsynek — APPROVED** [review]
LGTM, but I mostly checked the config code.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java [hunk 5]
```diff
@@ -43,12 +46,15 @@
...
         "scoreDrlList",
         "scoreDrlFileList",
+        "compileDroolsAlphaNetwork",
         "kieBaseConfigurationProperties",
         "initializingScoreTrend",
         "assertionScoreDirectorFactory"
 })
 public class ScoreDirectorFactoryConfig extends AbstractConfig<ScoreDirectorFactoryConfig> {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScoreDirectorFactoryConfig.class);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:56]: (separate issue)
in SolverManagerConfig we use the variable name `logger`, which violates name conventions, lets' fix this accross the code in a different PR?

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java (hunk 2 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:185]: In any case, let's not put it between the getters and setters, but before or after all the "simple getters and setters". These days we put the "simple getters and setters" at the end because it's just Java ceremony.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java [hunk 6]
```diff
@@ -123,6 +131,16 @@ public void setConstraintStreamImplType(ConstraintStreamImplType constraintStrea
         this.constraintStreamImplType = constraintStreamImplType;
     }
 
+    private boolean isUsingDrools() {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:134]: put before or after all simple getters/setters

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java [hunk 7]
```diff
@@ -123,6 +131,16 @@ public void setConstraintStreamImplType(ConstraintStreamImplType constraintStrea
...
+        if (scoreDrlList != null || scoreDrlFileList != null) { // We know we're in DRL.
+            return true;
+        } else if (constraintProviderClass == null) { // We know we're neither in DRL nor in CS.
+            return false;
+        }
+        // We know we're in CS.
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:140]: I'd remove this comment as it's obvious

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java [hunk 8]
```diff
@@ -156,6 +174,35 @@ public void setScoreDrlFileList(List<File> scoreDrlFileList) {
...
+        return compileDroolsAlphaNetwork;
+    }
+
+    public void setCompileDroolsAlphaNetwork(Boolean compileDroolsAlphaNetwork) {
+        this.compileDroolsAlphaNetwork = compileDroolsAlphaNetwork;
+    }
+
+    public boolean isDroolsAlphaNetworkCompilerEnabled() {
+        if (!isUsingDrools()) {
+            LOGGER.trace("Drools Alpha Network compiler is disabled when not using Drools.");
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:187]: An is***() method should not be doing logs like that. If it's used twice, we log twice.
A solvingStarted can - if desired - if the alpha network is enabled. But probably on a line along with other score director information (for example number of constraints) to avoid a verbose log.

See how much info the solvingSTarted line has

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java [hunk 9]
```diff
@@ -244,6 +291,11 @@ public ScoreDirectorFactoryConfig withScoreDrlFiles(File... scoreDrlFiles) {
         return this;
     }
 
+    public ScoreDirectorFactoryConfig withCompileDroolsAlphaNetworkEnabled(boolean compileDroolsAlphaNetwork) {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:294]: method name doesn't match field name "Enabled)

File: optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java [hunk 10]
```diff
@@ -70,6 +76,8 @@
     @XmlElement(name = "scoreDrlFile")
     protected List<File> scoreDrlFileList = null;
 
+    protected Boolean compileDroolsAlphaNetwork = null;
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:79]: Naming: maybe droolsAlphaNetworkCompilationEnabled ?

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/score/director/ScoreDirectorFactoryConfig.java:79]: reasoning: start with drools for context + easier to find if you were looking for something like this
end with Enabled to make it clear it's a switch and what true/false actually does (= turn it on/off)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java (hunk 3 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:191]: We typically call such methods "validate"

```suggestion
    private void validateNoAlphaNetworkCompiler() {
```

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:191]: Looks like this validation doesn't trigger if BAVET is used and it's true

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:191]: Maybe, you can just create a validateDroolsAlphaNetworkCompilation method that is always called, not just when it needs to be disabled?

**Comment by triceo** [comment]
Jenkins run native.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java (hunk 4 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java:41]: ConstraintStreamScoreDirectorFactory is not public API. Users don't know it exists.
So why do we need to split it up in Abstract* and subclasses?

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java:41]: ah, in the drools version we want to pass that boolean and in bavet it doesn't exist. ok, makes sense...

File: optaplanner-test/pom.xml [hunk 11]
```diff
@@ -64,6 +64,24 @@
         <groupId>org.revapi</groupId>
         <artifactId>revapi-maven-plugin</artifactId>
       </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-dependency-plugin</artifactId>
```
Comment (ge0ffrey) [inline optaplanner-test/pom.xml:69]: Is this temporary or permanent?

Comment (ge0ffrey) [reply optaplanner-test/pom.xml:69]: Let's add an expirary comment, something like add "Remove when we depend on drools directly" comment - or even better, with jira url to that task

Comment (ge0ffrey) [reply optaplanner-test/pom.xml:69]: => will be done in next pr anyway apparently, good to leave as is

File: optaplanner-test/src/main/java/org/optaplanner/test/api/score/stream/ConstraintVerifier.java [hunk 12]
```diff
@@ -60,6 +60,16 @@
...
             ConstraintStreamImplType constraintStreamImplType);
 
+    /**
+     * Applies only to {@link ConstraintStreamImplType#DROOLS}.
+     * Do not enable when running in a native image.
+     *
+     * @param droolsAlphaNetworkCompilerEnabled true to enable the alpha network compiler
+     * @return this
+     */
+    ConstraintVerifier<ConstraintProvider_, Solution_> withDroolsAlphaNetworkCompiler(
```
Comment (ge0ffrey) [inline optaplanner-test/src/main/java/org/optaplanner/test/api/score/stream/ConstraintVerifier.java:70]: if the field is called CompiLATION, this needs to be called ...ation too. And visa versa.

File: optaplanner-test/src/main/java/org/optaplanner/test/impl/score/stream/DefaultConstraintVerifier.java [hunk 13]
```diff
@@ -56,6 +56,21 @@ public ConstraintStreamImplType getConstraintStreamImplType() {
...
     }
 
+    public boolean isDroolsAlphaNetworkCompilerEnabled() {
+        return droolsAlphaNetworkCompilerEnabled;
+    }
+
+    @Override
+    public ConstraintVerifier<ConstraintProvider_, Solution_>
+            withDroolsAlphaNetworkCompiler(boolean droolsAlphaNetworkCompilerEnabled) {
+        if (getConstraintStreamImplType() != ConstraintStreamImplType.DROOLS && droolsAlphaNetworkCompilerEnabled) {
```
Comment (ge0ffrey) [inline optaplanner-test/src/main/java/org/optaplanner/test/impl/score/stream/DefaultConstraintVerifier.java:66]: I wouldn't even check the value of droolsAlphaNetworkCompilerEnabled.
If this is called and it's not using drools, let's just fail fast?

**Review by ge0ffrey — APPROVED** [review]
LGTM if naming is consistent and starts with drools + ends with Enabled

File: optaplanner-spring-integration/optaplanner-spring-boot-autoconfigure/src/main/java/org/optaplanner/spring/boot/autoconfigure/OptaPlannerAutoConfiguration.java [hunk 14]
```diff
@@ -179,11 +185,14 @@ public SolverConfig solverConfig() {
...
+                    solverConfig.getScoreDirectorFactoryConfig().isDroolsAlphaNetworkCompilerEnabled();
 
             return (ConstraintVerifier<ConstraintProvider_, SolutionClass_>) ConstraintVerifier.build(constraintProvider,
-                    solverConfig.getSolutionClass(),
-                    entityClasses).withConstraintStreamImplType(
-                            (constraintStreamImplType != null) ? constraintStreamImplType : ConstraintStreamImplType.DROOLS);
+                    solverConfig.getSolutionClass(), entityClasses)
+                    .withConstraintStreamImplType(
+                            (constraintStreamImplType != null) ? constraintStreamImplType : ConstraintStreamImplType.DROOLS)
+                    .withDroolsAlphaNetworkCompiler(droolsAlphaNetworkCompilerEnabled);
```
Comment (ge0ffrey) [inline optaplanner-spring-integration/optaplanner-spring-boot-autoconfigure/src/main/java/org/optaplanner/spring/boot/autoconfigure/OptaPlannerAutoConfiguration.java:195]: ah, that's why the validation accepts false.
Maybe it should accept Boolean? Idunno. As is is fine too. This is serious nitpick. Feel free to ignore.

File: optaplanner-spring-integration/optaplanner-spring-boot-autoconfigure/src/main/java/org/optaplanner/spring/boot/autoconfigure/OptaPlannerAutoConfiguration.java [hunk 15]
```diff
@@ -162,6 +162,12 @@ public SolverConfig solverConfig() {
                         throw new UnsupportedOperationException(noConstraintProviderErrorMsg);
                     }
 
+                    @Override
+                    public ConstraintVerifier<ConstraintProvider_, SolutionClass_> withDroolsAlphaNetworkCompiler(
+                            boolean droolsAlphaNetworkCompilerEnabled) {
+                        throw new UnsupportedOperationException(noConstraintProviderErrorMsg);
```
Comment (ge0ffrey) [inline optaplanner-spring-integration/optaplanner-spring-boot-autoconfigure/src/main/java/org/optaplanner/spring/boot/autoconfigure/OptaPlannerAutoConfiguration.java:168]: (separate issue, not part of this PR) cc @Christopher-Chianelli if you could take a look, Chris:

Let's also improve the error message for clarity, so it :
```
final String noConstraintProviderErrorMsg = "ConstraintVerifier is only supported for ConstraintProviders";
```
=>
```
final String noConstraintProviderErrorMsg = "Cannot create a " + ConstraintVerifier.class.getSimpleName() + " because there is no " + ConstraintProvider.class.getSimpleName() + " configured in the " + SolverConfig.class.getSimpleName() + ".";
```

**Comment by triceo** [comment]
Jenkins run native.

**Review by Christopher-Chianelli — APPROVED** [review]
LGTM

**Comment by triceo** [comment]
Jenkins run native.
